### ISSUE TITLE
iconv should have the encoding ISO646-SE2

### ISSUE THREAD

**Issue by mobluse** [git-for-windows/git#1655]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?
Git-2.16.2-64-bit.exe

```
$ git --version --build-options

git version 2.16.2.windows.1
cpu: x86_64
built from commit: e1848984d1004040ec5199e749b5f282ddf4bb09
sizeof-long: 4
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.17128.1]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Nano
Path Option: CmdTools
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
curl -s https://www.abc.se/programbanken/abc/abc80/asmkod/basicii.txt | dos2unix | iconv -f ISO646-SE2 -t UTF8 | less
```
 - What did you expect to occur after running these commands?

The file should be shown in `less`.

 - What actually happened instead?

A blank file in less, and after quit I saw these error messages:
```
iconv: conversion from ISO646-SE2 unsupported
iconv: try 'iconv -l' to get the list of supported encodings
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

** insert URL here **

I installed from `Git-2.16.2-64-bit.exe` from git-scm.com. `iconv` is included in this package. I think `iconv` should have the encoding `ISO646-SE2`. Ubuntu 16.04 has this encoding. I use it to read old Swedish text files and programs, which there are a lot of e.g.:
`curl -s https://www.abc.se/programbanken/abc/abc80/asmkod/basicii.txt | dos2unix | iconv -f ISO646-SE2 -t UTF8 | less`
`ISO646-SE2` is used by e.g. the retro-computers Luxor [ABC80](https://en.wikipedia.org/wiki/ABC_80) (1978) and [ABC800](https://en.wikipedia.org/wiki/ABC_800)-series (1981). At my university we only have Git Bash and not Ubuntu for WSL on Windows 10 computers.

**Comment by dscho** [comment]
Here is how you can find out from where the problem stems:

- install the [Git for Windows SDK](https://gitforwindows.org/#download-sdk)
- run `git-bash.exe` inside the cloned worktree
- run `pacman -Qo $(which iconv.exe)`

If the package name starts with `mingw-w64-`, it is maintained in https://github.com/Alexpux/MINGW-packages (probably in the `mingw-w64-iconv/` subdirectory), otherwise in https://github.com/Alexpux/MSYS2-packages (it would be in the `iconv/` subdirectory, then).

You could report the bug there, but you could do even better: try to fix it. For that, first rebuild the package locally, as described here: https://github.com/git-for-windows/git/wiki/Package-management#rebuild-packages. You will then have the source code in `src/` and the full description how it was built in the `build()` function in `PKGBUILD`.

**Comment by dscho** [comment]
> pacman was not found.

Wait, waaaat?

Does `usr\bin\pacman.exe` exist in your clone? Did you start `git-bash.exe` in `C:\Users\pi\Downloads\git-sdk` or your regular Git Bash (the latter won't find `pacman.exe` for obvious reasons...)?

**Comment by dscho** [comment]
> I suppose this means that I can report the issue to https://github.com/Alexpux/MSYS2-packages/tree/master/libiconv.
> 
> I searched for ISO646- in that repository, but it could not be found.

The MSYS2-packages repository only contains build scripts. You really have to follow [my earlier suggestion](https://github.com/git-for-windows/git/issues/1655#issuecomment-384575503):

> For that, first rebuild the package locally, as described here: https://github.com/git-for-windows/git/wiki/Package-management#rebuild-packages. You will then have the source code in src/ and the full description how it was built in the build() function in PKGBUILD.

> The use of pacman indicates that this package comes from Arch Linux.

Not at all. Pacman originates from Arch Linux, this is true, but it is also used as package manager of MSYS2. I would be utterly surprised if MSYS2's `iconv` package and Arch Linux' one had anything more in common than the name and the unpatched source code.

**Comment by drizzd** [comment]
It seems that while MSYS2's iconv comes from http://www.gnu.org/software/libiconv/, Ubuntu's iconv comes from glibc: https://askubuntu.com/a/898835.

**Comment by mobluse** [comment]
`iconv` in Arch Linux has [ISO646-SE2](https://en.wikipedia.org/wiki/ISO/IEC_646#cite_ref-52).

I discovered that `piconv` exists in Git Bash, but that also doesn't have ISO646-SE2.

A workaround is to use 
```
tr '`{|}~@[\\]^$\177' 'éäöåüÉÄÖÅÜ¤■'
```
_e.g._
```
 curl -s https://www.abc.se/programbanken/abc/abc800/spel/adventur/starwars.bas | tr '`{|}~@[\\]^$\177' 'éäöåüÉÄÖÅÜ¤■' | dos2unix | less
```

**Comment by drizzd** [comment]
I tried to compile glibc's iconv in msys2, but it is strongly coupled to glibc's build system, which of course is ill prepared to run in msys2, or on Windows. But if someone knows glibc better, maybe it can be done.

```
git clone http://sourceware.org/git/glibc.git
cd glibc
# Avoid "checking sysdep dirs... configure: error: Operating system msys is not supported."
cp -a sysdeps/x86_64 sysdeps/msys
mkdir build
cd build
../configure --enable-hacker-mode
make iconv/subdir_lib
```

Output:

```
[...]
gcc iconv_open.c -c -std=gnu11 -fgnu89-inline  -O2 -Wall -Werror -Wundef -Wwrite-strings -fmerge-all-constants -fno-stack-protector -frounding-math -g -Wstrict-prototypes -Wold-style-definition -fno-math-errno      -ftls-model=initial-exec      -I../include -I/c/Users/Ricl/src/glibc/build/iconv  -I/c/Users/Ricl/src/glibc/build  -I../sysdeps/x86_64/64  -I../sysdeps/x86_64/fpu  -I../sysdeps/x86/fpu/include -I../sysdeps/x86/fpu  -I../sysdeps/x86_64  -I../sysdeps/x86  -I../sysdeps/ieee754/float128  -I../sysdeps/ieee754/ldbl-96/include -I../sysdeps/ieee754/ldbl-96  -I../sysdeps/ieee754/dbl-64/wordsize-64  -I../sysdeps/ieee754/dbl-64  -I../sysdeps/ieee754/flt-32  -I../sysdeps/msys  -I../sysdeps/wordsize-64  -I../sysdeps/ieee754  -I../sysdeps/generic  -I.. -I../libio -I.   -include /c/Users/Ricl/src/glibc/build/libc-modules.h -DMODULE_NAME=libc -include ../include/libc-symbols.h  -DPIC     -DTOP_NAMESPACE=glibc -o /c/Users/Ricl/src/glibc/build/iconv/iconv_open.o -MD -MP -MF /c/Users/Ricl/src/glibc/build/iconv/iconv_open.o.dt -MT /c/Users/Ricl/src/glibc/build/iconv/iconv_open.o
In file included from <command-line>:0:0:
./../include/libc-symbols.h:34:32: error: "MODULE_libc" is not defined, evaluates to 0 [-Werror=undef]
 #define IN_MODULE  PASTE_NAME (MODULE_, MODULE_NAME)
                                ^
./../include/libc-symbols.h:32:26: note: in definition of macro 'PASTE_NAME1'
 #define PASTE_NAME1(a,b) a##b
                          ^
./../include/libc-symbols.h:34:20: note: in expansion of macro 'PASTE_NAME'
 #define IN_MODULE  PASTE_NAME (MODULE_, MODULE_NAME)
                    ^~~~~~~~~~
./../include/libc-symbols.h:35:22: note: in expansion of macro 'IN_MODULE'
 #define IS_IN(lib)  (IN_MODULE == MODULE_##lib)
                      ^~~~~~~~~
./../include/libc-symbols.h:47:5: note: in expansion of macro 'IS_IN'
 #if IS_IN (testsuite) || defined IS_IN_build || defined __cplusplus
     ^~~~~
```

**Comment by dscho** [comment]
@mobluse and to be precise: you can use [this commit](http://git.savannah.gnu.org/cgit/libiconv.git/commit/?id=3a33986e1d2c819dc2b8a84684e067cf177d1815) which added support for ISO-2022-JP-MS as a blueprint to add support for ISO646-SE1. I am sure you can do it, you now have all the tools that you need in your hands.

**Comment by mobluse** [comment]
@dscho Well, it looks rather complicated to add support for ISO646-SE2 in libiconv. I think libiconv doesn't support other 7-bit encodings other than plain ASCII. Maybe they (GNU) have a policy of not accepting 7-bit encodings into the code-base, because they don't interfere with UTF-8 control codes (e.g. ISO-2022-JP-MS seems to be an 8-bit extended ASCII). In that case we would have to maintain our own fork of libiconv. Also, ISO646-SE2 is not enough for ABC80 and ABC800 I discovered yesterday since \\x7F (delete) is a printing character that also can be typed on the ABC80-keyboard (Ctrl+<) and thus is used in texts and source code. This delete character is not even in UTF-8 AFAIK -- it is a standing, filled rectangle the size of an 'M'. I approximated it so far with a square '■', but that is really too low. The Swedish ABC80 used Mullard SAA5052 for character generation, but I now discovered that they probably also used the other [Mullard SAA5050](https://en.wikipedia.org/wiki/Mullard_SAA5050)-type chips for other markets.
